Feature: add abacus toolchain

QuantumMisaka opened the pull request:
Toolchain scripts for quickly and properly compile and install dependencies of ABACUS including:

- gnu-toolchain
- - `gcc` (always in system)
- - `openblas`
- - `openmpi` or `mpich`
- - `scalapack`, `fftw`, `elpa` , `libxc`
- - `cereal`, `libnpy`, `libtorch`
- intel-toolchain (likely)

Version 2023.1 inspired by cp2k-toolchain.
Still need more test and improvement, but can have good usage.

The toolchain directory is designed to be in the path of ABACUS itself.

QuantumMisaka committed:
Toolkits: abacus toolchain

Toolchain scripts for quickly and properly compile and install dependencies of ABACUS including:

- gnu-toolchain
- - `gcc` (always in system)
- - `openblas`
- - `openmpi` or `mpich`
- - `scalapack`, `fftw`, `elpa` , `libxc`
- - `cereal`, `libnpy`, `libtorch`
- intel-toolchain (likely)

Version 2023.1 inspired by cp2k-toolchain.
Still need more test and improvement, but can have good usage.

The toolchain directory is designed to be in the path of ABACUS itself.

dyzheng commented:
There are too much new files in this PR. Please check your changes.

QuantumMisaka committed:
make file lesser

too much file mainly in libs/cereal and libs/libnpy, make them easier so less is more

QuantumMisaka commented:
@dyzheng  Got it, much new file mainly in `libs/cereal` and `libs/libnpy`, now fix it.

caic99 reviewed (changes requested):
@QuantumMisaka 
Thanks for your contribution! I believe these script will help users a lot.

However, please remove the release packages under libs. Doing so is an action of  redistribution and integration of those software packages. This is strongly not recommended because it might introduce incompatibility on licenses. (And you did not include the license files of those packages.)

It is OK to download them in your script. If you think integrating them in ABACUS repo is necessary, please use git submodule.
Thanks!

caic99 commented:
@QuantumMisaka And, if you wish, you can add docs or links to the toolchain under corresponding parts of `abacus-develop/docs` in online docs to promote it.

QuantumMisaka commented:
@caic99 I also want to remove the libraries in `libs` and use toolchain to install it, I'm trying to do this but It's hard to choose the way to introduce the installation (by `git clone` or `wget` from the release, and neither way can bypass `github.com`, which will bring the HTTP problem). So I highly want to solve this with your help.

Also, the `docs` file is in plan. I'm wandering if the README file can be directly copy to `docs` directory, 

Thanks !

caic99 commented:
> @caic99 I also want to remove the libraries in `libs` and use toolchain to install it, I'm trying to do this but It's hard to choose the way to introduce the installation (by `git clone` or `wget` from the release, and neither way can bypass `github.com`, which will bring the HTTP problem). So I highly want to solve this with your help.

In principle it is the user's responsibiliy guaranteeing the internet connection. However, considering the actual situations, maybe you can provide some mirror sites of GitHub (e.g. fastgit) in place of the fixed base domain of github.com . Using [gitee mirroring service](https://gitee.com/mirrors) is a more secure way. 

> Also, the `docs` file is in plan. I'm wandering if the README file can be directly copy to `docs` directory

Please choose a good place for it to maintain the structure. Maybe under [quick_start](https://github.com/deepmodeling/abacus-develop/tree/develop/docs/quick_start) folder, and add an introduction for it at [easy_install.md](https://github.com/deepmodeling/abacus-develop/blob/develop/docs/quick_start/easy_install.md)? What do you think?

QuantumMisaka committed:
Merge branch 'develop' into develop

wenfei-li commented on toolchain/libs/cereal-1.3.2.tar.gz:
I'm wondering if it is appropriate to include tar.gz file in the ABACUS repository

QuantumMisaka replied:
It's surely not, I'm trying to add `cereal` and `libnpy` in toolchain installation

QuantumMisaka replied:
Fixed it

QuantumMisaka committed:
version 2023.2

- use toolchain scripts to install `cereal` and `libnpy` by `wget` from github.com, remove `libs` directory at the same time
- fix some bug in intel-toolchain
- update README
- update executable permission

QuantumMisaka committed:
Merge branch 'develop' into develop

QuantumMisaka committed:
update README

update readme

QuantumMisaka committed:
Merge branch 'develop' of https://github.com/QuantumMisaka/abacus-develop into develop

caic99 commented on toolchain/README.md:
```diff
@@ -0,0 +1,109 @@
+- [x] `intel-mkl-mpi` toolchain support (for old oneapi using `icc` instead of `icx`).
+- [x] `intelmkl-openmpi/mpich` toolchain support (need more test).
+- [x] Automatic installation of [CEREAL](https://github.com/USCiLab/cereal) and [LIBNPY](https://github.com/llohse/libnpy) (by github.com)
+- [ ] A better mirror station for all packages, espeacially for CEREAL and LIBNPY.
```
```suggestion
- [ ] A better mirror station for all packages, especially for CEREAL and LIBNPY.
```

QuantumMisaka committed:
Update toolchain/README.md

Co-authored-by: Chun Cai <amoycaic@gmail.com>

QuantumMisaka committed:
Update docs

- update README
- update easy_install.md

QuantumMisaka committed:
Merge branch 'develop' of https://github.com/QuantumMisaka/abacus-develop into develop

QuantumMisaka committed:
Update README.md

add target for acml toolchain

QuantumMisaka committed:
Update easy_install.md

QuantumMisaka committed:
update intel-mpich toolchain

update a quick way to use intel-mpich toolchain installation
which has the best speed in test on intel-8358 HPC server

QuantumMisaka committed:
Merge branch 'develop' into develop

QuantumMisaka committed:
minor fixed

QuantumMisaka committed:
Merge branch 'develop' of https://github.com/QuantumMisaka/abacus-develop into develop

QuantumMisaka committed:
minor fixed and tag a problem

QuantumMisaka committed:
Update toolchain_gnu.sh

default gnu toolchain change to openmpi ,which is more efficient in most of the test

QuantumMisaka committed:
Merge branch 'deepmodeling:develop' into develop

QuantumMisaka committed:
version 2023.3

- update env setting after installation in `setup`
- update build method of intel-toolchain and intel-mpich toolchain
- minor bug fixed

QuantumMisaka committed:
Merge branch 'develop' into develop

QuantumMisaka committed:
Update easy_install.md

QuantumMisaka committed:
minor fixed

QuantumMisaka committed:
Update toolchain output

QuantumMisaka committed:
FIx: Update gcc setup

QuantumMisaka committed:
FIx: executable permission

QuantumMisaka committed:
Update: CEREAL & LIBNPY installation

use a more normal way to introduce cereal & libnpy

QuantumMisaka committed:
Change: leave place for LibRI Support

QuantumMisaka commented:
@caic99 A problem in my try for introducing `LibRI` in `dep` to toolchain installation by using intel-icc-toolchain
```bash
In file included from /lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Tensor.h(8),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_lcao/hamilt_lcaodft/LCAO_matrix.h(13),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_io/restart.h(6),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_pw/hamilt_pwdft/global.h(8),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp(1):
/lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Shared_Vector.h(8): catastrophic error: cannot open source file "cereal/cereal.hpp"
  #include <cereal/cereal.hpp>

compilation aborted for /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp (code 4)
gmake[2]: *** [source/module_cell/CMakeFiles/cell.dir/build.make:244: source/module_cell/CMakeFiles/cell.dir/klist.cpp.o] Error 4
```
What's the best way for `LibRI` to find `CEREAL` ?

QuantumMisaka committed:
Merge branch 'develop' into develop

caic99 commented:
> @caic99 A problem in my try for introducing `LibRI` in `dep` to toolchain installation by using intel-icc-toolchain
> 
> ```shell
> In file included from /lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Tensor.h(8),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_lcao/hamilt_lcaodft/LCAO_matrix.h(13),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_io/restart.h(6),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_pw/hamilt_pwdft/global.h(8),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp(1):
> /lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Shared_Vector.h(8): catastrophic error: cannot open source file "cereal/cereal.hpp"
>   #include <cereal/cereal.hpp>
> 
> compilation aborted for /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp (code 4)
> gmake[2]: *** [source/module_cell/CMakeFiles/cell.dir/build.make:244: source/module_cell/CMakeFiles/cell.dir/klist.cpp.o] Error 4
> ```
> 
> What's the best way for `LibRI` to find `CEREAL` ?

Hi @QuantumMisaka ,
Please check if `LibRI` has `CEREAL` for its git submodule. If not so, you can add `CEREAL` as an submodule in your toolchain.

caic99 commented:
> Hi @QuantumMisaka , Please check if `LibRI` has `CEREAL` for its git submodule. If not so, you can add `CEREAL` as an submodule in your toolchain.

@QuantumMisaka Update: Please check in your terminal if `cereal/cereal.hpp` can be opened in editor like `vim`. If the file is not present, the warning message would be `fatal error: xxx.h: No such file or directory`, but not `catastrophic error: cannot open source file`. The latter means the file presents, but the compiler can not open it. Please check file access permission.

QuantumMisaka committed:
Update install_openmpi.sh

QuantumMisaka committed:
Merge branch 'develop' of https://github.com/QuantumMisaka/abacus-develop into develop

QuantumMisaka commented:
> @QuantumMisaka Update: Please check in your terminal if `cereal/cereal.hpp` can be opened in editor like `vim`. If the file is not present, the warning message would be `fatal error: xxx.h: No such file or directory`, but not `catastrophic error: cannot open source file`. The latter means the file presents, but the compiler can not open it. Please check file access permission.

- The permission of `cereal.hpp` is 644, which can be open by `vim`, 
- The problem still exists after change permission of `cereal.hpp` to `777`
- The problem will be fixed if `export CPATH="cereal/include":$CPATH

QuantumMisaka committed:
Fix: Support LibRI installation

QuantumMisaka committed:
Fix: Downgrade default libtorch to version 1.12.1

Libtorch 2.0.1 will lead to lots of warning in abacus build process

QuantumMisaka committed:
minoe fixed
